Molecular module

SunTzunami opened the pull request:
A module for reading molecule data which can be then used to create actors.

pep8speaks commented:
Hello @SunTzunami! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`fury/molecular.py`](https://github.com/fury-gl/fury/blob/67b1ad8da2e7a3c67a7ef5b34249ef37417a585b/fury/molecular.py):

> [Line 5:1](https://github.com/fury-gl/fury/blob/67b1ad8da2e7a3c67a7ef5b34249ef37417a585b/fury/molecular.py#L5): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 154:80](https://github.com/fury-gl/fury/blob/67b1ad8da2e7a3c67a7ef5b34249ef37417a585b/fury/molecular.py#L154): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)

SunTzunami commented on fury/molecular.py:
```diff
@@ -0,0 +1,483 @@
+            elif multipleBonds == 'Off':
+                self.setMultiTubesForBonds(False)
+        else:
+            print("Inadequate Bonding data")
```
For better error handling, should I make a custom exception? 
Something like -
```
class BondingError(Exception):
 pass
```
and then use it by calling 
`raise BondingError("Inadequate Bonding Error!")`

SunTzunami replied:
Done

SunTzunami commented on fury/molecular.py:
```diff
@@ -0,0 +1,600 @@
+        """
+        return self.GetDefaultRGBTuple(atomicNumber)
+
+
```
The functions below this comment are used to create actors for the molecules by using MoleculeMapper.
Will it be more appropriate to have them in the molecular module or should they be added to actor module?

SunTzunami commented on fury/molecular.py:
```diff
@@ -0,0 +1,187 @@
+        """
+        return self.GetAtomPosition(atom_index)
+
+    def setAtomicPosition(self, atom_index, atom_coordinate):
```
Thank you @nasimanousheh ! Nice catch!

skoudoro commented:
Hi @SunTzunami, Do not forget to update the coding style in this PR. Let us know when it is ready for review. 

Thanks!

SunTzunami commented:
Hi @skoudoro ! Thanks for the reminder! 
 I tried to make it more pythonic. With the exception of [PeriodicTable class](https://github.com/fury-gl/fury/pull/452/files#diff-8cedb8d1d2aca3f6f359201c5623b21f1f9b55329c4e7cae44d396e9a62bdec5R519), no other class has member functions. 
I'm guessing that this class will probably remain untouched throughout further releases as it functions as a kind of a library which can be used to obtain properties of elements and this data remains fairly constant for long time (until a new element is discovered or experimental data prove the old data wrong which rarely happers). **If needed I can remove its member functions too.** 
I also added an example to show how to parse a PDBx file and create the three molecular visualizations. 
@nasimanousheh @guaje 
PTAL.

skoudoro commented on docs/examples/demo_molecular.py:
```diff
@@ -0,0 +1,113 @@
+###############################################################################
+# Parsing the mmCIF file for information about coordinates and atoms
+
+pdbxfile = open(pdbxfn, 'r')
+pdbx_lines = pdbxfile.readlines()
+for line in pdbx_lines:
+    _line = line.split()
+    try:
+        if _line[0] == 'ATOM' or _line[0] == 'HETATM':
+
+            # obtain coordinates of atom
+            coorX, coorY, coorZ = float(_line[10]), float(_line[11]), \
+                                  float(_line[12])
+
+            # obtain the atomic number of atom
+            atomic_num = table.get_atomic_number(_line[2])
+
+            # add the atomic data to the molecule
+            molecular.add_atom(molecule, atomic_num, coorX, coorY, coorZ)
+
+    except Exception:
+        continue
```
Why do you need to do it manually? I believe the `molecular` module will handle this part. otherwise, you need to create a reader in `fury.io`

skoudoro commented on docs/examples/demo_molecular.py:
```diff
@@ -0,0 +1,113 @@
+
+
+# stick representation
+scene.add(molecular.molecular_stick_rep_actor(molecule, bond_thickness=2))
```
`molecular.molecular....` : this is repetitive, the actor need to change name

skoudoro commented on docs/examples/demo_molecular.py:
```diff
@@ -0,0 +1,113 @@
+# stick representation
+scene.add(molecular.molecular_stick_rep_actor(molecule, bond_thickness=2))
+
+# ball and stick representation
+# scene.add(molecular.molecular_bstick_rep_actor(molecule,
+#                                                atom_scale_factor=0.3,
+#                                                bond_thickness=2))
+
+# sphere representation
+# scene.add(molecular.molecular_sphere_rep_actor(molecule))
```
if not use, comment to remove

SunTzunami replied:
These can be un-commented to see the other models. I added this example for the members to understand how to use the molecular module.
I'll create the final, proper demo for the module once the module has been made properly (in a pythonic way).

skoudoro replied:
ok, sounds good

skoudoro commented on docs/examples/demo_molecular.py:
```diff
@@ -0,0 +1,113 @@
+
+###############################################################################
+# Delete the PDBx file if it's downloaded from the internet
+if flag:
+    os.remove(outfnm)
```
not necessary if you use the fetcher

skoudoro commented on fury/molecular.py:
```diff
@@ -0,0 +1,729 @@
+    molecule.AppendBond(atom1_index, atom2_index, bond_type)
+
+
+def get_total_num_atoms(molecule):
```
as said before, this should be a property.   In general, if you need a get/set, please, use property tag to be more pythonic

skoudoro commented on fury/molecular.py:
```diff
@@ -0,0 +1,729 @@
+    return molecule.GetNumberOfAtoms()
+
+
+def get_total_num_bonds(molecule):
```
same comment as above

skoudoro commented on fury/molecular.py:
```diff
@@ -0,0 +1,729 @@
+    return molecule.GetNumberOfBonds()
+
+
+def get_atomic_number(molecule, atom_index):
```
same comment as above

skoudoro commented on fury/molecular.py:
```diff
@@ -0,0 +1,729 @@
+    return molecule.GetAtomAtomicNumber(atom_index)
+
+
+def set_atomic_number(molecule, atom_index, atomic_num):
```
same comment as above

skoudoro commented on fury/molecular.py:
```diff
@@ -0,0 +1,729 @@
+    deep_copy(molecule, bonder.GetOutput())
+
+
+class MoleculeMapper(vtk.vtkOpenGLMoleculeMapper):
+    """Class to create mappers for three types of molecular represenations-
+    1. Ball and Stick Representation.
+    2. Stick Representation.
+    3. Sphere Representation.
+
+    It is used to create mappers which are then used to create actors for the
+    above mentioned representations.
+    """
```
What is the goal of this empty class?

SunTzunami replied:
It's used to create the mappers which are then used to create actors. Please see [here](https://github.com/fury-gl/fury/pull/452/files#diff-8cedb8d1d2aca3f6f359201c5623b21f1f9b55329c4e7cae44d396e9a62bdec5R632) to see how it's used.

skoudoro replied:
It is strange to have an empty class. I do not think you need it. 

I should use directly vtkOpenglMoleculeMapper in your different function

SunTzunami replied:
Okay! I'll use vtkOpenGLMoleculeMapper directly.

skoudoro reviewed (commented):
Hi @SunTzunami,

Thank you for the update. See below for some comments. Good progress but it still needs some work. Can you start to add some tests?

Thank you

SunTzunami commented on fury/molecular.py:
```diff
@@ -0,0 +1,729 @@
+        Actor created to render the ball and stick representation of the
+        molecule to be visualized.
+    """
+    bs_mapper = MoleculeMapper()
```
@skoudoro  the MoleculeMapper class is used here to create the mapper for the ball stick actor.

SunTzunami commented on fury/molecular.py:
```diff
@@ -0,0 +1,729 @@
+        Actor created to render the stick representation of the molecule to be
+        visualized.
+    """
+    mst_mapper = MoleculeMapper()
```
@skoudoro  the MoleculeMapper class is used here to create the mapper for stick actor.

SunTzunami commented on fury/molecular.py:
```diff
@@ -0,0 +1,729 @@
+        Actor created to render the space filling representation of the
+        molecule to be visualized.
+    """
+    msp_mapper = MoleculeMapper()
```
@skoudoro  the MoleculeMapper class is used here to create the mapper for sphere representation actor.

SunTzunami commented:
Hi @skoudoro ! As per your suggestions, I updated the module to remove the empty class and used `@property` in the [PeriodicTable class](https://github.com/fury-gl/fury/pull/452/files#diff-8cedb8d1d2aca3f6f359201c5623b21f1f9b55329c4e7cae44d396e9a62bdec5R254) to make it more pythonic.

SunTzunami commented:
Hi @skoudoro ! For Molecule class, 
I tried using the following type of code and was able to make "get" type of member functions to work - 
```
@property
def total_num_atoms(self):
"""Returns the total number of atoms in a given molecule."""
return self.GetNumberOfAtoms()
```
For "set" type of member functions, I tried to use following type of code but it doesn't work -
```
@atomic_number.setter
def atomic_number(self, atom_index, atomic_num):
    """Set the atomic number of an atom for a specified index.

    Assign atomic_num as the atomic number of the atom present at
    atom_index.

    Parameters
    ----------
    atom_index : int
        Index of the atom to whom the atomic number is to be assigned.
    atom_num : int
        Atomic number to be assigned to the atom.
    """
    self.SetAtomAtomicNumber(atom_index, atomic_num)
```
This gives an error `'function' object has no attribute 'setter'` as atomic_number is not a data member of the Molecule class. Any suggestions for what should be done for the "get" type of functions?

skoudoro commented:
I have not seen the paramater before. You can look at this example: https://stackoverflow.com/a/5718537 and it would be fine

SunTzunami commented:
Hi @skoudoro ! Thanks for the example, I tried implementing it, 
```
## for getting
def atomic_number(self, atom_index):
    """Get the atomic number of an atom for a specified index.

    Returns the atomic number of the atom present at index atom_index.

    Parameters
    ----------
    atom_index : int
        Index of the atom whose atomic number is to be obtained.
    """
    return self.GetAtomAtomicNumber(atom_index)

## for setting
def set_atomic_number(self, atom_index, atomic_num):
    """Set the atomic number of an atom for a specified index.

    Assign atomic_num as the atomic number of the atom present at
    atom_index.

    Parameters
    ----------
    atom_index : int
        Index of the atom to whom the atomic number is to be assigned.
    atom_num : int
        Atomic number to be assigned to the atom.
    """
    self.SetAtomAtomicNumber(atom_index, atomic_num)
atomic_number = property(set_atomic_number)
```
It overrides the getting member function when I do this.

SunTzunami commented:
Hi @skoudoro ! I'm done with the tests, PTAL.

nasimanousheh commented on fury/molecular.py:
```diff
@@ -0,0 +1,698 @@
+        The molecule to which the bond belongs.
+    bond_index : int
+        Index of the atom to which the coordinates are to be assigned.
+    bond_type : int
```
bond_type should be string not integer.

SunTzunami replied:
changed bond_type to bond_order

nasimanousheh commented on fury/molecular.py:
```diff
@@ -0,0 +1,698 @@
+    molecule.SetAtomPosition(atom_index, x_coord, y_coord, z_coord)
+
+
+def get_bond_type(molecule, bond_index):
```
Please rename "bond_type" to "bond_order", Bond type is different than bond order. A data file can have different bond types with same bond order. For example, the bond type of C-C is different than than bond type of C-H but both of them have bond order of 1.

SunTzunami replied:
Thanks! Done.

nasimanousheh commented on fury/molecular.py:
```diff
@@ -0,0 +1,698 @@
+            * 'VDW' : for Van Der Waals radius of the atom
+            * 'Covalent' : for covalent radius of the atom
+        """
+        if radius_type == 'VDW':
```
Do you know which one ('VDW' or 'Covalent') is by default? 
You can raise an error if radius_type is something else.

SunTzunami replied:
Post comment, I made it such that it's 'VDW' by default and a value error is raised if it's something other than these two. Thanks!

SunTzunami commented on docs/examples/demo_molecular.py:
```diff
@@ -0,0 +1,113 @@
+###############################################################################
+# Doing 5 things here -
+# 1. Creating a scene object
+# 2. configuring the camera's position
```
done

SunTzunami commented on docs/examples/demo_molecular.py:
```diff
@@ -0,0 +1,113 @@
+# 2. configuring the camera's position
+# 3. Creating and adding axes actor to the scene
+# 4. Computing the bonding information for the molecule.
+# 4. Generating and adding molecular model to the scene.
```
done

SunTzunami commented:
Hi @nasimanousheh @skoudoro ! I've made changes as suggested in the meetings. PTAL.

SunTzunami commented on fury/molecular.py:
```diff
@@ -0,0 +1,704 @@
+    elif colormode == 'single':
+        msp_mapper.SetAtomColorMode(0)
+    else:
+        msp_mapper.SetAtomColorMode(1)
```
Done

skoudoro commented:
Thank you for all this update @SunTzunami. I will into it asap. 

@nasimanousheh, can you do a first review? Thanks !

skoudoro commented on fury/molecular.py:
```diff
@@ -0,0 +1,704 @@
+    for i in range(num_total_atoms):
+        secondary_structures[i] = ord('c')
+        resi = molecule.residue_seq[i]
+        for j in range(len(molecule.sheet)):
+            sheet = molecule.sheet[j]
```
use enumerate instead

SunTzunami replied:
done

skoudoro commented on fury/molecular.py:
```diff
@@ -0,0 +1,704 @@
+                continue
+            secondary_structures[i] = ord('s')
+
+        for j in range(len(molecule.helix)):
+            helix = molecule.helix[j]
```
use enumerate instead, more pythonic

SunTzunami replied:
done

skoudoro reviewed (commented):
Hi @SunTzunami,

Thank you for the update! see below for some minor comment. It is almost ready to go. We are just waiting for @nasimanousheh review

SunTzunami commented on fury/molecular.py:
```diff
@@ -0,0 +1,704 @@
+        helix : ndarray of integers, shape (H, 4) where H is the total number
+            of helices present in the molecule.
+            Array containing information about helices present in the molecule.
+        is_hetatm : ndarray of bools
```
done

SunTzunami commented on fury/molecular.py:
```diff
@@ -0,0 +1,704 @@
+            Atomic number of the element whose covalent radius is to be
+            obtained.
+        radius_type : string
+            Two valid choices -
```
added colon instead

SunTzunami commented on fury/molecular.py:
```diff
@@ -0,0 +1,704 @@
+    ----------
+    molecule : Molecule() object
+        The molecule to be rendered.
+    colormode : string
```
done

SunTzunami commented on fury/molecular.py:
```diff
@@ -0,0 +1,704 @@
+          multiple tubes.
+        * 'off': all bonds (single, double, triple) will be shown as single
+          bonds (i.e shown using one tube each).
+        Default is 'on'.
```
done

SunTzunami commented on fury/molecular.py:
```diff
@@ -0,0 +1,704 @@
+    bond_thickness : float
+        Used to manipulate the thickness of bonds (i.e. thickness of tubes
+        which are used to render bonds).
+        Default is 0.1. (Optimal range: 0.1 - 0.5)
```
done

SunTzunami commented:
Thanks for the review @skoudoro ! I've addressed all the comments.

SunTzunami commented:
Hi @skoudoro ! Keeping #457 in mind, I was able to use partial imports by using following snippet:
```python
from vtkmodules.all import VTK_ID_TYPE, VTK_UNSIGNED_SHORT, \
                           VTK_UNSIGNED_CHAR, VTK_UNSIGNED_INT,\
                           vtkDataSetAttributes, vtkActor, vtkPolyData, \
                           vtkStringArray, vtkUnsignedCharArray, \
                           vtkFloatArray, vtkMolecule, \
                           vtkSimpleBondPerceiver, vtkPeriodicTable, \
                           vtkProteinRibbonFilter, vtkOpenGLMoleculeMapper, \
                           vtkPolyDataMapper
```
It works and even the tests work.
Should I push the changes or do it in a future PR?

skoudoro commented:
Hi @SunTzunami,

Thank you for anticipating that.

You should avoid `from vtkmodules.all`  but use instead `from vtkmodules.vtkCommonCore import `.

Using `from vtkmodules.all` is like doing `import vtk`. However, we want to limit the number of modules imported. You need to be more specific.

You can look at my PR [here](https://github.com/fury-gl/fury/pull/470/files#diff-0987f453f29064de86e45c8cda10256f2334dbfd8c52145e682816b9e62a4e1eR10)

SunTzunami commented:
Hi @skoudoro ! This worked for me (tests all passed too):

```python
import vtkmodules.vtkCommonCore as ccvtk
import vtkmodules.vtkCommonDataModel as cdcvtk
import vtkmodules.vtkRenderingCore as rcvtk
import vtkmodules.vtkDomainsChemistry as dcvtk
import vtkmodules.vtkDomainsChemistryOpenGL2 as dcovtk
```
Is this the correct way? Should I push the changes?

nasimanousheh reviewed (commented):
Thank you @SunTzunami for updating your PR. 
Please find my comments.

SunTzunami commented on fury/molecular.py:
```diff
@@ -0,0 +1,716 @@
+    """
+    bonder = vtk.vtkSimpleBondPerceiver()
+    bonder.SetInputData(molecule)
+    bonder.SetTolerance(0.1)
```
Basically, vtkSimpleBondPerceiver performs a simple check of all interatomic distances and adds a single bond between atoms that are reasonably close. If the interatomic distance is less than the sum of the two atom's covalent radiiplus a tolerance (mathematically, (d < r1 + r2 + tolerance)), a single bond is added. 
By default, vtkSimpleBondPerceiver has a high tolerance of 0.45, I thought that this default tolerance is quite high and might lead to excessive bonding and hence chose an arbitrary value of 0.1.

SunTzunami commented on fury/molecular.py:
```diff
@@ -0,0 +1,716 @@
+    return vtk_to_numpy(molecule.GetAtomicPositionArray().GetData())
+
+
+def deep_copy_molecule(molecule1, molecule2):
```
It's not possible to transfer molecular information from molecule2 intro molecule1 by simple assignment i.e. by setting molecule1 = molecule2. Deep copy copies the molecular information from molecule2 into molecule1. You can check the vtk documentation [here](https://vtk.org/doc/nightly/html/classvtkMolecule.html#a227d15dc1266116895d65e72319f273b).

SunTzunami commented on fury/tests/test_molecular.py:
```diff
@@ -0,0 +1,336 @@
+    # Test errors for inadequate bonding data
+    npt.assert_raises(ValueError, molecular.ball_stick, molecule)
+
+    molecular.add_bond(molecule, 0, 1, 1)
```
To answer the first part, yes it's correct, as you can see on [this](https://github.com/fury-gl/fury/pull/452/files#diff-7c3fdef0e37ff1e81acaf4ff0275d29a6b65c070226c5721dd7b5a0fd54c52bcR186) line, `multiple_bonds` is set to "Off" for the second window and as a result it'll show only single bonds for double/triple bonds.
for `bond_order` = 4, I'll create a note in the documentation.

SunTzunami commented on docs/experimental/viz_molecular_demo.py:
```diff
@@ -0,0 +1,163 @@
+A small example to show how to use the space_filling_model actor and
+and generate a space filling model for a protein. This example also
+shows how to parse PDBx/mmCIF files to obtain atomic info essential
+(coordinates and element names) to constructing the model.
```
done

SunTzunami commented on docs/experimental/viz_molecular_demo.py:
```diff
@@ -0,0 +1,163 @@
+        print("Error in downloading the file!")
+
+###############################################################################
+# creating an PeriodicTable() object to obtain atomic numbers from names of
```
done

SunTzunami commented on fury/molecular.py:
```diff
@@ -0,0 +1,716 @@
+    molecule.AppendAtom(atomic_num, x_coord, y_coord, z_coord)
+
+
+def add_bond(molecule, atom1_index, atom2_index, bond_order=1):
```
I created a note

SunTzunami commented on fury/molecular.py:
```diff
@@ -0,0 +1,716 @@
+    molecule : Molecule() object
+        The molecule to which the bond belongs.
+    bond_index : int
+        Index of the atom to which the coordinates are to be assigned.
```
it should be "index of the bond". Thank you!

SunTzunami commented:
Hi @nasimanousheh ! Thanks for the feedback. I've addressed your comments, PTAL
As for ``add_bond``,  when the ``bond_order`` exceeds 4, this leads to erratic behaviour and I have mentioned that this should be avoided in the function's notes.

skoudoro commented:
> Is this the correct way? Should I push the changes?

yes! just replace `import vtkmodules.vtkCommonDataModel as cdcvtk`  by `import vtkmodules.vtkCommonDataModel as cdmvtk`

SunTzunami commented:
Thank you @skoudoro ! I made the replacement and used partial imports.

SunTzunami commented:
added a function to create a bounding box, Preview: 

![4ury](https://user-images.githubusercontent.com/65067354/128624529-03c026be-7f80-4792-b57e-eceeb1767ec2.png)

I don't know why the test for bounding box is failing on azure pipeline. It's passing on my Linux (ubuntu 20.04). [Link](https://dev.azure.com/fury-gl/fury/_build/results?buildId=1761&view=logs&j=dcd50a69-7bd8-59f7-72b1-a119e6ffb32f&t=44a0de46-c7b4-5cae-8a63-c46b346a7582&l=156) to the azure error message.

SunTzunami commented on fury/tests/test_molecular.py:
```diff
@@ -0,0 +1,359 @@
+
+    molecule_actor = mol.stick(molecule)
+    test_box = mol.bounding_box(molecule)
+    """
+    scene.add(molecule_actor)
+    scene.add(test_box)
+
+    if interactive:
+        window.show(scene)
+
+    npt.assert_equal(scene.GetActors().GetNumberOfItems(), 2)
+
+    table = mol.PeriodicTable()
+    colors = np.array([table.atom_color(1), table.atom_color(6)])
+    arr = window.snapshot(scene)
+    report = window.analyze_snapshot(arr, colors=colors)
+    npt.assert_equal(report.objects, 1)
+    scene.clear()
+    """
```
For some reason, the tests pass only when the actor is not added to the scene. This is strange as the tests pass locally for me without commenting out this section.

Garyfallidis commented on docs/experimental/viz_molecular_demo.py:
```diff
@@ -0,0 +1,164 @@
+"""
+======================================================================
+Space filling model actor
```
Can you make the title more specific to molecules?

Garyfallidis replied:
Also should we cite some papers? There must be some original papers that promoted these representations.

SunTzunami replied:
Good catch! The description was for an earlier version of the demo. I updated the title and description.

skoudoro replied:
This file is in experimental, should we move it to tutorial @Garyfallidis @SunTzunami @nasimanousheh ?

SunTzunami replied:
I don't think we should move this to tutorial as this was just for the reviewers/mentors to understand how to work with the module. 
I've made proper tutorials (using `fetcher`) for which I'll create PRs once this PR is merged.

Garyfallidis commented on docs/experimental/viz_molecular_demo.py:
```diff
@@ -0,0 +1,164 @@
+if flag:
+    os.remove(outfnm)
+
+interactive = True
```
False please.

SunTzunami replied:
Done

Garyfallidis commented on fury/molecular.py:
```diff
@@ -0,0 +1,749 @@
+
+def ball_stick(molecule, colormode='discrete',
+               atom_scale_factor=0.3, bond_thickness=0.1,
+               multiple_bonds='on'):
```
Why not just use a boolean here? True/False

SunTzunami replied:
Done. Makes sense. Nice suggestion!

Garyfallidis commented on fury/molecular.py:
```diff
@@ -0,0 +1,749 @@
+    multiple_bonds : string, optional
+        Set whether multiple tubes will be used to represent multiple
+        bonds. Two valid choices:
+        * 'on': multiple bonds (double, triple) will be shown by using
```
Same as above.

SunTzunami replied:
Done.

Garyfallidis commented:
Also, the bounding box doesn't look good. Can you make something nicer using ``fury.actor.streamtubes`` ?

Garyfallidis commented:
And the png up says 4ury and not fury if that was the intended name.

SunTzunami commented:
Hi @Garyfallidis ! Thanks for the review. bounding box is using `actor.streamtubes` now.  
Preview: 

![image](https://user-images.githubusercontent.com/65067354/130282136-8ccdb6a9-1616-4d35-b75e-bd868e870a7b.png)

Yes, it's supposed to say **4ury** as that is indeed the protein's name (click [here](https://www.rcsb.org/structure/4URY) for more info about the protein).  I made it all caps now.
I thought it was a nice play on words as the protein has a name which is pretty similar to the library's name.

SunTzunami commented:
Hi @Garyfallidis ! 
I found three papers which were the first ones to adopt/create these representations (from the wikipedia pages of the respective representations). I couldn't find any papers for stick representation. 
Also, these papers are kinda old (published in 1971, 1952, 1981 respectively).

For Ball and Stick: https://pubs.acs.org/doi/abs/10.1021/ed048p407
For Space-filling: https://aip.scitation.org/doi/10.1063/1.1770803
For Ribbon: https://www.sciencedirect.com/science/article/abs/pii/S0065323308605203

SunTzunami commented:
Hi @Garyfallidis ! PTAL.

Garyfallidis commented:
It is fine if the citations are old. No problem with that @SunTzunami . Please add the references to the docstrings of the relevant functions and tutorial.

SunTzunami commented:
Hi @Garyfallidis  I've added the references. PTAL.

Garyfallidis commented:
Hi @SunTzunami, can you rebase this? There should be more CIs passing after.

Garyfallidis commented on docs/experimental/viz_molecular_demo.py:
```diff
@@ -0,0 +1,184 @@
+======================================================================
+
+A small example to show how to use the various molecular
+representations present in the `molecular` module to visualize
```
representations present?
Change verb please.

SunTzunami replied:
done

Garyfallidis commented on docs/experimental/viz_molecular_demo.py:
```diff
@@ -0,0 +1,184 @@
+###############################################################################
+# Downloading the PDB file of the protein to be rendered.
+# User can change the pdb_code depending on which protein they want to
+# visualize.
```
blank space missing...

SunTzunami replied:
I changed the description

Garyfallidis commented on docs/experimental/viz_molecular_demo.py:
```diff
@@ -0,0 +1,184 @@
+sheets = np.array(sheets)
+helix = np.array(helix)
+is_hetatm = np.array(is_hetatm)
+
```
too many empty lines

SunTzunami replied:
removed the unnecessary ones

Garyfallidis reviewed (changes requested):
A few small things left @SunTzunami please make the changes and update the continuous integration files to the master. Apologies, if you have done that already and I somehow missed it. But do let us know about your status.

SunTzunami commented:
Hi @Garyfallidis 
Thanks for the review, I've rebased my PR and made the corrections to the docstring as suggested. PTAL

Garyfallidis commented:
Great work @SunTzunami ! Merging ... !
